### PR TITLE
Add Dockerfile for arm64 build container

### PR DESCRIPTION
**Pull request by xoloki**
## Description

This change adds a `Dockerfile` that creates a build container for the `arm64` architecture.  There is an absolute nightmare of dependencies for this repository, so having a container will save everyone time (myself included).

## Type of Change
- New feature

## Does this introduce a breaking change?
No

## Are documentation updates required?
No

### CODE REVIEW

File: Dockerfile
Comment (djordon) [inline Dockerfile:32]: Could this be added to the first `RUN apt-get install` as well.

File: Dockerfile
Comment (djordon) [inline Dockerfile:41]: Is this necessary if cargo is already installed?

File: Dockerfile
Comment (djordon) [inline Dockerfile:31]: What does this do?

**Comment by AshtonStephens** [comment]
Looking

File: Dockerfile
Comment (xoloki) [reply Dockerfile]: Done.

File: Dockerfile
Comment (xoloki) [reply Dockerfile:1]: It's considered a best practice to only use the new ubuntu LTS after the August point release updates.  Before that the LTS is not exactly stable.

File: DOCKER.md
Comment (xoloki) [reply DOCKER.md:1]: Done.

File: DOCKER.md
Comment (xoloki) [reply DOCKER.md]: Done.

File: Dockerfile
Comment (xoloki) [reply Dockerfile]: I'm really not a fan of multi-line docker commands.  

I do however agree on cleaning up.

File: Dockerfile
Comment (xoloki) [reply Dockerfile]: Done.

File: Dockerfile
Comment (xoloki) [reply Dockerfile:27]: This one has to be after the `nodesource` install because that adds repositories.

File: Dockerfile
Comment (xoloki) [reply Dockerfile:36]: It's necessary to be able to get the correct version of `rustc`.

File: Dockerfile
Comment (djordon) [reply Dockerfile:1]: Oh yeah that makes perfect sense.

File: Dockerfile
Comment (djordon) [reply Dockerfile:27]: Oh right, that makes sense. The `https://deb.nodesource.com` url should have been a giveaway.

File: Dockerfile
Comment (djordon) [reply Dockerfile:36]: Really? That's surprising, I thought it would auto-update to the correct version after running `cargo build`.

Comment (xoloki) [reply Dockerfile:36]: It was complaining about having `rustc` `1.75` but needing `1.76` or more recent.

File: Dockerfile
Comment (AshtonStephens) [inline Dockerfile:3]:
```suggestion
MAINTAINER Joey Yandle <joey@trustmachines.co>
```

File: Dockerfile
Comment (djordon) [reply Dockerfile:36]: Oh right, that's because apt-get doesn't install `rustup`, which I suppose cargo delegates to when it notices that it needs a different version of rust. Okay, never mind.

Comment (AshtonStephens) [reply Dockerfile:36]: I prefer the all in one `curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs/ | sh -s -- -y` so just want to make sure you considered it and are choosing a different path. I prefer this because there's no artifact; we aren't checking any hash here so we don't really need one.

File: README.md
Comment (AshtonStephens) [inline README.md:112]: For my own education, can you elaborate on why this works? The Makefile hasn't received any edits so I don't understand how this would do what it does.

File: Dockerfile
Comment (djordon) [reply Dockerfile:3]: `MAINTAINER` is [deprecated](https://docs.docker.com/reference/dockerfile/#label), but I think this should be changed to
```docker
LABEL org.opencontainers.image.authors="Joey Yandle <joey@trustmachines.co>"
```

File: Dockerfile
Comment (djordon) [reply Dockerfile]: Fair enough. Also we'd have to explicitly add `ca-certificates` for my suggestion to work.

File: Dockerfile
Comment (xoloki) [reply Dockerfile]: Fixed @djordon @AshtonStephens

**Comment by xoloki** [comment]
@djordon @AshtonStephens when you're good please resolve convos and if necessary approve separately.

**Comment by AshtonStephens** [comment]
Approving, I think this looks good and is better than everyone doing everything locally. I think there might be some rough edges with us having two build setups now but we can fix that down the line. I've opened this ticket for it: https://github.com/stacks-network/sbtc/issues/236

**Review by AshtonStephens — APPROVED** [review]

**Review by AshtonStephens — CHANGES_REQUESTED** [review]
Actually sorry, getting this error:

```cli
Packages: +649
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 ERR_PNPM_ENOENT  ENOENT: no such file or directory, copyfile '/build/.pnpm-store/v3/files/7f/65c6403a23d93fb148e8259b012d6552ab3bff178f4a7d6a9d9cec0f60429fc1899e39b4bca8cc08afc75d9a7c7bfdb13fc372ca63c85eb22b0355eb4d6000' -> '/build/node_modules/.pnpm/@nestjs+core@10.3.0_@nestjs+common@10.3.0_reflect-metadata@0.1.13_rxjs@7.8.1__reflect-metadata@0.1.13_rxjs@7.8.1/node_modules/@nestjs/core_tmp_17/inspector/interfaces/extras.interface.js'
```

Any ideas?

**Comment by xoloki** [comment]
> Actually sorry, getting this error:
> 
> ```
> Packages: +649
> ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
>  ERR_PNPM_ENOENT  ENOENT: no such file or directory, copyfile '/build/.pnpm-store/v3/files/7f/65c6403a23d93fb148e8259b012d6552ab3bff178f4a7d6a9d9cec0f60429fc1899e39b4bca8cc08afc75d9a7c7bfdb13fc372ca63c85eb22b0355eb4d6000' -> '/build/node_modules/.pnpm/@nestjs+core@10.3.0_@nestjs+common@10.3.0_reflect-metadata@0.1.13_rxjs@7.8.1__reflect-metadata@0.1.13_rxjs@7.8.1/node_modules/@nestjs/core_tmp_17/inspector/interfaces/extras.interface.js'
> ```
> 
> Any ideas?
> 
> And separately this when I try to run `make build` in the container.
> 
> ```
> > blocklist-api@ build-openapi /build/.generated-sources/blocklist-openapi-gen
> > openapi-generator-cli generate -i ./blocklist-client-openapi.json -g rust -o ../blocklist-api --config ./config.json
> 
> sh: 1: openapi-generator-cli: not found
>  ELIFECYCLE  Command failed.
> ```

Did you run `make install` before `make build` @AshtonStephens ?

**Comment by xoloki** [comment]
You might also need a fresh checkout @AshtonStephens since mounting `pwd` means that the built code in on the host not the container.

**Comment by AshtonStephens** [comment]
> Did you run `make install` before `make build` @AshtonStephens ?

Yes the first error is from the make install. Okay, I'll try a fresh instance of the repository.

**Comment by xoloki** [comment]
> > Did you run `make install` before `make build` @AshtonStephens ?
> 
> Yes the first error is from the make install. Okay, I'll try a fresh instance of the repository.

Just tried fresh checkout of source repo and `make install` def succeeded.  Running `make build` now

**Comment by AshtonStephens** [comment]
I think, though, that there are some rough edges here that need to be rounded out, so I'd keep this method of building the repository out of the `README.md` until we've had a chance to make this a bit more robust. But in the immediate term it makes sense to unblock your development.

**Comment by xoloki** [comment]
> I think, though, that there are some rough edges here that need to be rounded out, so I'd keep this method of building the repository out of the `README.md` until we've had a chance to make this a bit more robust. But in the immediate term it makes sense to unblock your development.

I'll move it back to `DOCKER.md`

**Comment by xoloki** [comment]
Ok back to separate `DOCKER.md` @AshtonStephens @djordon

**Comment by AshtonStephens** [comment]
@xoloki why don't we merge this tomorrow, I might be the weak link here but for whatever reason this isn't working for me and I don't want to rush this in and have a confusing build setup that only works on a subset of computers.

**Comment by xoloki** [comment]
Ok @AshtonStephens happy to merge whenever I get a +1.  Thanks for helping me get this out the door.

File: .screenrc
Comment (djordon) [inline .screenrc:1]: I think I overlooked this yesterday, but why is this needed in the repo? I would think that it's better to gitignore custom configuration files like this in cases where someone else prefers a different setup. We can still mount a local version during `docker run`. Thoughts?

Comment (xoloki) [reply .screenrc:1]: I always `.gitignore` `*~` so emacs backup files don't get added accidentally @djordon , but I'll remove it from this PR since it's orthogonal to the change.

**Review by AshtonStephens — APPROVED** [review]
Approving pending @djordon's approval.

File: .screenrc
Comment (xoloki) [reply .screenrc:1]: Done.

Comment (djordon) [reply .screenrc:1]: I think you removed the `.gitignore` file, but I want to add the `.screenrc` file to the `.gitignore` file (and remove copying it to the `Dockerfile`). This way you can still have and use it locally by mounting your local copy to the right place during `docker run` and others can use their preferred configuration files as well.

**Review by djordon — APPROVED** [review]
Looks good ✅

**Comment by djordon** [comment]
Just to be clear, I'm also getting the same errors as @AshtonStephens but we're going to fix this up later right?

**Pull request merged**
